runners/google-cloud-dataflow-java/worker/src/test/java/org/apache/beam/runners/dataflow/worker/StreamingDataflowWorkerTest.java

```diff
@@ -3046,19 +3048,25 @@ public void testHugeCommits() throws Exception {
 
   private static class SlowDoFn extends DoFn<String, String> {
 
+    private final Duration sleep;
+
+    SlowDoFn(Duration sleep) {
+      this.sleep = sleep;
+    }
+
+    SlowDoFn() {
+      this(Duration.millis(1000));
+    }
+
     @ProcessElement
     public void processElement(ProcessContext c) throws Exception {
-      Thread.sleep(1000);
+      Uninterruptibles.sleepUninterruptibly(sleep.getMillis(), TimeUnit.MILLISECONDS);
```

Comment:
would be better to inject a clock to this and the same one into StreamingDataflowWorker instead of relying on sleep timing

Can you take a look at if that would be not too bad to add? Somewhat existing but would be nice to cleanup while modifying

Reply:
Injecting a fake clock works, but the unit tests still depend on the wall clock via Timer or ScheduledExecutorService. A fake ScheduledExecutorService is work in progress.

Reply:
This pull request now has another commit that injects a fake ScheduledExecutorService, making the new unit tests completely independent from the wall-time clock.